src/main/java/net/dv8tion/jda/internal/handle/GuildMemberRemoveHandler.java

```diff
@@ -55,27 +58,47 @@ protected Long handleInternally(DataObject content)
             //We probably just left the guild and this event is trying to remove us from the guild, therefore ignore
             return null;
         }
+
+        // Update the memberCount
+        guild.onMemberRemove();
+
+        User user = api.getEntityBuilder().createFakeUser(content.getObject("user"));
         MemberImpl member = (MemberImpl) guild.getMembersView().remove(userId);
 
         if (member == null)
         {
             WebSocketClient.LOG.debug("Received GUILD_MEMBER_REMOVE for a Member that does not exist in the specified Guild. UserId: {} GuildId: {}", userId, id);
+            // Remove user from voice channel if applicable
+            guild.getVoiceChannelsView().forEachUnordered((channel) -> {
+                VoiceChannelImpl impl = (VoiceChannelImpl) channel;
+                Member connected = impl.getConnectedMembersMap().remove(userId);
```

Comment:
If we don't handle the voice states we just have an empty map. Then `connected` is null and the following if check will evaluate to false.

Reply:
Right, my definition of fail was "not emit an event when we should".
From your statement it seems that if we aren't caching members we don't have their voice states and they don't appear in the connectedMembersMap. As such, there _wouldnt_ be an event to fire, right?

We can resolve this if that's the case.

Reply:
If voice states are tracked, they will be in the `conntectedMembers` map but not necessarily in the guild's member cache. Voice state cache and member cache are not connected that strongly. Since the member will be in this map we can still fire the event.

Reply:
Does that mean that this code should run a check against the cache to try and provide a more up-to-date member before returning?
https://github.com/DV8FromTheWorld/JDA/blob/feature/gateway-intents/src/main/java/net/dv8tion/jda/internal/entities/VoiceChannelImpl.java#L71

Reply:
https://github.com/DV8FromTheWorld/JDA/blob/75c9b392a33fb9956d3c591590dc7292b54074a1/src/main/java/net/dv8tion/jda/internal/entities/VoiceChannelImpl.java#L139-L147

Reply:
I _swear_ I looked at that method better. I'd even spoken about the real vs member in a different comment. Idk how I didn't see the `getConnectedMembersMap` call instead of just `connectedMembersMap`.